File: solutions/medium/142.Linked_List_Cycle_II/142.Linked_List_Cycle_II.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        nodeSeen = set()
        while head != None:
            if head in nodeSeen:
                return head
            nodeSeen.add(head)
            head = head.next
        return None

File: solutions/medium/142.Linked_List_Cycle_II/142.Linked_List_Cycle_II.ts

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

function detectCycle(head: ListNode | null): ListNode | null {
  let nodeSeen = new Set()
  while (head !== null) {
    if (nodeSeen.has(head)) return head
    nodeSeen.add(head)
    head = head.next
  }
  return null;
};